patches/tModLoader/Terraria/ModLoader/EmoteBubbleLoader.cs

```diff
@@ -0,0 +1,174 @@
+﻿using Microsoft.Xna.Framework;
+using Microsoft.Xna.Framework.Graphics;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Terraria.GameContent.UI;
+using Terraria.GameContent.UI.Elements;
+using Terraria.Localization;
+
+namespace Terraria.ModLoader
+{
+	public static class EmoteBubbleLoader
+	{
+		public static int EmoteBubbleCount => emoteBubbles.Count + EmoteID.Count;
+		internal static readonly List<ModEmoteBubble> emoteBubbles = new();
+		internal static readonly List<GlobalEmoteBubble> globalEmoteBubbles = new();
+		internal static readonly Dictionary<int, List<ModEmoteBubble>> categoryEmoteLookup = new();
+
+		internal static int Add(ModEmoteBubble emoteBubble) {
+			if (ModNet.AllowVanillaClients)
+				throw new Exception("Adding emote bubbles breaks vanilla client compatibility");
+
+			emoteBubbles.Add(emoteBubble);
+			return EmoteBubbleCount - 1;
+		}
+
+		internal static void Unload() {
+			emoteBubbles.Clear();
+			globalEmoteBubbles.Clear();
+			categoryEmoteLookup.Clear();
+		}
+		
+		internal static void ResizeArrays() {
+			Array.Resize(ref Lang._emojiNameCache, EmoteBubbleCount);
+
+			for (int k = EmoteID.Count; k < EmoteBubbleCount; k++) {
+				Lang._emojiNameCache[k] = LocalizedText.Empty;
+			}
+		}
+
+		internal static Dictionary<Mod, List<int>> GetAllUnlockedModEmotes() {
+			var result = new Dictionary<Mod, List<int>>();
+			foreach (var modEmoteBubble in from modEmote in emoteBubbles where modEmote.IsUnlocked() select modEmote)
```

Comment:
Please use 'functional style' LINQ in tML. 
```cs
foreach (var modEmote in emoteBubbles.Where(e => e.IsUnlocked())
```

Alternatively, since it's just a simple filter, put it in the loop:
```cs
foreach (var modEmote in emoteBubbles) {
    if (!modeEmote.isUnlocked()) continue;
    ...
```